lib/blackbeard/context.rb

```diff
@@ -53,15 +53,26 @@ def unique_identifier
     end
 
     def user_id
+      unless bbd.nil?
+        VisitorUserTracker.set_visitor_id_for_user(user_id: @user.id, visitor_id: bbd)
+      end
+
+      vid = VisitorUserTracker.get_visitor_id_for_user(@user.id)
+      return vid if vid
       @user.id
     end
 
     def visitor_id
-      @visitor_id ||= (controller.request.cookies['bbd'] || generate_visitor_id).to_i
+      @visitor_id ||= (bbd || generate_visitor_id).to_i
     end
 
 private
 
+    def bbd
+      return nil if controller.nil? || controller.request.nil?
+      controller.request.cookies['bbd']
```

Comment:
We would need to use `controller.try(:request).try(:cookies).try(:[], "bbd").to_i` to get rid of the conditional.